### PR TITLE
Restructure training functionality

### PR DESCRIPTION
**Pull request by rhine3**
Restructure training functionality to separate loading audio data in a dataset from training using that dataset.

Changes to `opensoundscape/datasets.py` are all related to the class `BinaryFromAudio`, which I renamed to `SingleTargetAudioDataset`. I made changes to other files to make this renaming consistent. Major changes within `datasets.py` include:
* Making `SingleTargetAudioDataset` return a consistent shape and removing its `spec_augment` argument
* Trimming using audio trim, not spectrogram trim, and fixing bugs related to trimming length
* Changing in-pipeline augmentation (random trimming and overlaying) to a methods-based format in order to make `__getitem__` easier to interpret

Major changes to `opensoundscape/torch/train.py` include:
* Removing calls to `SingleTargetAudioDataset`
* Changing some arguments to accept datasets instead of dataframes
* Changing the way that spectrogram augmentation is handled, with different reshaping consistent with changes made to `SingleTargetAudioDataset`

### CODE REVIEW

File: opensoundscape/datasets.py [hunk 1]
```diff
@@ -178,138 +179,174 @@ def collate_fn(*batch):
...
+        # apply desired random transformations to image and convert to tensor
         X = self.transform(image)
 
-        if self.spec_augment:
-            X = X[0].unsqueeze(0)
-
+        # Return data : label pairs (training/validation)
         if self.label_column:
-            labels = np.array([row[self.label_column]])
+            labels = np.array([self.row[self.label_column]])
```
Comment (chiroptical) [inline opensoundscape/datasets.py:348]: Why save this in the Dataset?

File: opensoundscape/torch/train.py [hunk 2]
```diff
@@ -81,9 +80,14 @@ def train(
             targets = y.squeeze(1)
 
             if spec_augment:
+                # X is currently shape [1, 3, width, height]
+                # Take to shape [1, 1, width, height] for use with `augment`
+                X = X[0][0].unsqueeze(0).unsqueeze(0)
```
Comment (chiroptical) [inline opensoundscape/torch/train.py:85]: Is this right? I read this as take the first image in the batch only.

Comment (chiroptical) [reply opensoundscape/torch/train.py:85]: Seems like `X[:][0].unsqueeze(0)`?

File: tests/test_datasets.py [hunk 3]
```diff
@@ -120,28 +120,29 @@ def test_basic_splitting_operation_with_include_last_segment(
...
-    dataset = BinaryFromAudio(
-        binary_from_audio_df,
+def test_single_target_audio_dataset_tensor_augment(single_target_audio_dataset_df):
+    dataset = SingleTargetAudioDataset(
+        single_target_audio_dataset_df,
         label_column="NumericLabels",
         height=225,
         width=226,
-        spec_augment=True,
+        tensor_augment=True,
```
Comment (chiroptical) [inline tests/test_datasets.py:137]: Shouldn't this fail?

File: opensoundscape/datasets.py [hunk 4]
```diff
@@ -178,138 +179,174 @@ def collate_fn(*batch):
...
+    def random_audio_trim(self, audio, audio_length):
+        audio_length = len(audio.samples) / audio.sample_rate
+        if self.random_trim_length > audio_length:
+            raise ValueError(
+                f"the length of the original file ({audio_length} sec) was less than the length to extract ({self.random_trim_length} sec) for the file {audio_path}"
+            )
+        extra_time = audio_length - self.random_trim_length
+        start_time = np.random.uniform() * extra_time
+        trimmed = audio.trim(start_time, start_time + self.random_trim_length)
+        return trimmed, self.random_trim_length
```
Comment (chiroptical) [inline opensoundscape/datasets.py:268]: return of `self.audio_trim_length` is probably unnecessary here

Comment (rhine3) [reply opensoundscape/datasets.py]: Fixed by https://github.com/kitzeslab/opensoundscape/pull/179/commits/bbfccac7e1e6eb9f73583f861c0720a2c1f45cc7

File: opensoundscape/datasets.py (hunk 1 above)
Comment (rhine3) [reply opensoundscape/datasets.py]: Fixed by https://github.com/kitzeslab/opensoundscape/pull/179/commits/bbfccac7e1e6eb9f73583f861c0720a2c1f45cc7

File: opensoundscape/torch/train.py (hunk 2 above)
Comment (rhine3) [reply opensoundscape/torch/train.py]: Fixed by https://github.com/kitzeslab/opensoundscape/pull/179/commits/1ea680bd155ff542b622389947df723254c17c98

File: tests/test_datasets.py (hunk 3 above)
Comment (rhine3) [reply tests/test_datasets.py]: All tests related to this PR now pass

**Pull request merged**